### PR TITLE
os/: st7701 LCD driver with MIPI

### PR DESCRIPTION
**Pull request by namanjain7**

### CODE REVIEW

File: os/drivers/Makefile [hunk 1]
```diff
@@ -100,6 +100,8 @@ include usbdev$(DELIM)Make.defs
 include usbhost$(DELIM)Make.defs
 include video$(DELIM)Make.defs
 include wireless$(DELIM)Make.defs
+include mipidsi$(DELIM)Make.defs
```
Comment (sunghan-chang) [inline os/drivers/Makefile:103]: Let's add this by Alphabet order.

File: os/board/rtl8730e/src/rtl8730e_st7701.c [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	u8 green;
+	u8 red;
+	u8 alpha;
+} argb8888_t;
+
+typedef struct {
+	u8 blue;
+	u8 green;
+	u8 red;
+} rgb888_t;
```
Comment (sunghan-chang) [inline os/board/rtl8730e/src/rtl8730e_st7701.c:60]: They look common structures, not st7701 specific. How about moving them to other common file? Because other file also uses this and defines there duplicately.

File: os/board/rtl8730e/src/rtl8730e_st7701.c [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+static int rtl8730_st7701_lcd_enable(void);
+static int rtl8730_st7701_lcd_layer_enable(int layer, bool enable);
+static void lcd_put_area(u8 *LcdImgBuffer, u32 x1, u32 y1, u32 x2, u32 y2);
+static void LcdcEnable(void);
+
+struct st7701_config_s g_rtl8730_st7701_config_s = {
+	.init = rtl8730_st7701_lcd_init,
+	.reset = rtl8730_st7701_gpio_reset,
+    .cache_invalidate = rtl8730_st7701_cache_invalidate,
+    .lcd_reload = rtl8730_st7701_lcd_reload,
```
Comment (sunghan-chang) [inline os/board/rtl8730e/src/rtl8730e_st7701.c:43]: Different indentation. Let's use a tab.

File: os/board/rtl8730e/src/rtl8730e_st7701.c [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+LCDC_IRQInfo LcdcIrqInfo = {
+    .IrqNum = LCDC_IRQ,
+    .IrqPriority = INT_PRI_MIDDLE,
+    .IrqData = (u32)LCDC,
+};
+
+LCDC_TypeDef *pLCDC = LCDC;
+LCDC_InitTypeDef LCDC_InitStruct;
+
+u8 *LcdImgBuffer1 = (u8 *)(DDR_BASE + LCDC_IMG_BUF_OFFSET2 + LCDC_IMG_BUF_SIZE);
```
Comment (sunghan-chang) [inline os/board/rtl8730e/src/rtl8730e_st7701.c:79]: In our coding rules, we don't allow High / Low character mixing. Let's use like `lcd_img_buffer1`.
Same as other variables, functions and so on.

File: os/board/rtl8730e/src/rtl8730e_st7701.c [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	rtl8730_st7701_lcd_init();
+
+	if (lcddev_register(dev) < 0) {
+		lldbg("ERROR: LCD driver register fail\n");
+	} else {
+		lldbg("LCD driver register success\n");
+	}
+	LcdcEnable();
+	return;
+}
```
Comment (sunghan-chang) [inline os/board/rtl8730e/src/rtl8730e_st7701.c:202]: Could you check the warning `no line end`?

File: os/drivers/lcd/lcd_dev.c [hunk 6]
```diff
@@ -68,7 +68,6 @@
 #include <tinyara/fs/ioctl.h>
 #include <tinyara/lcd/lcd_dev.h>
 
-#define MAX_NO_PLANES 3
```
Comment (sunghan-chang) [inline os/drivers/lcd/lcd_dev.c:71]: This definition is used at some files. Could you let me know where you define this instead of here?

File: os/drivers/lcd/st7701.c [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+ *  dev     - The lcd device
+ *  row     - Starting row to read from (range: 0 <= row < yres)
+ *  col     - Starting column to read read (range: 0 <= col <= xres-npixels)
+ *  buffer  - The buffer in which to return the run read from the LCD
+ *  npixels - The number of pixels to read from the LCD
+ *            (range: 0 < npixels <= xres-col)
+ *
+ ****************************************************************************/
+
+#ifndef CONFIG_LCD_NOGETRUN
```
Comment (sunghan-chang) [inline os/drivers/lcd/st7701.c:239]: To work this line, you should include `<tinyara/config.h>`.

File: os/drivers/mipidsi/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations under
+# the License.
+#
+# ##############################################################################
+
+if(CONFIG_MIPI_DSI)
```
Comment (sunghan-chang) [inline os/drivers/mipidsi/CMakeLists.txt:21]: Is this file necessary?

File: os/drivers/mipidsi/Kconfig [hunk 9]
```diff
@@ -0,0 +1,19 @@
+#
+# For a description of the syntax of this configuration file,
+# see the file kconfig-language.txt in the NuttX tools repository.
```
Comment (sunghan-chang) [inline os/drivers/mipidsi/Kconfig:3]: Let's use our sentence. Please find other Kconfig file.

File: os/drivers/mipidsi/Make.defs [hunk 10]
```diff
@@ -0,0 +1,35 @@
+############################################################################
+# drivers/video/mipidsi/Make.defs
```
Comment (sunghan-chang) [inline os/drivers/mipidsi/Make.defs:2]: 1. Need to update the path.
2. Need to add the Samsung license.

File: os/drivers/mipidsi/Make.defs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ifeq ($(CONFIG_MIPI_DSI),y)
+
+CSRCS += mipi_dsi_device.c mipi_dsi_host.c mipi_dsi_packet.c mipi_dsi_device_driver.c mipi_dsi_host_driver.c
+
+ifeq ($(CONFIG_MIPI_DSI_DRIVER),y)
+	#CSRCS += panel_st7701.c
+endif
+
+DEPPATH += --dep-path mipidsi
+#CFLAGS += ${INCDIR_PREFIX}$(TOPDIR)$(DELIM)drivers$(DELIM)mipidsi
```
Comment (sunghan-chang) [inline os/drivers/mipidsi/Make.defs:32]: If unnecessary, let's remove the line

File: os/drivers/mipidsi/mipi_dsi.h [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+ * Public Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Functions Definitions
+ ****************************************************************************/
+
+#ifdef __cplusplus
+#define EXTERN extern "C"
+extern "C"
```
Comment (sunghan-chang) [inline os/drivers/mipidsi/mipi_dsi.h:52]: extern "C" {

File: os/drivers/mipidsi/mipi_dsi_device.c [hunk 13]
```diff
@@ -0,0 +1,972 @@
...
+ ****************************************************************************/
+
+ssize_t mipi_dsi_transfer(FAR struct mipi_dsi_device *device, FAR struct mipi_dsi_msg *msg)
+{
+	FAR const struct mipi_dsi_host_ops *ops = device->host->ops;
+
+	if (ops == NULL || ops->transfer == NULL) {
+		return -ENOSYS;
+	}
+	printf("mipi transfer is used\n");
```
Comment (sunghan-chang) [inline os/drivers/mipidsi/mipi_dsi_device.c:80]: Is this necessary? If yes, we can change this to xxxdbg.

File: os/drivers/mipidsi/mipi_dsi_device.c [hunk 14]
```diff
@@ -0,0 +1,972 @@
...
+	/*if ((device->mode_flags & MIPI_DSI_MODE_LPM) != 0)
```
Comment (sunghan-chang) [inline os/drivers/mipidsi/mipi_dsi_device.c:81]: Do we need to keep commented codes?

File: os/include/tinyara/mipidsi/mipi_dsi.h [hunk 15]
```diff
@@ -0,0 +1,878 @@
...
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+struct mipi_dsi_host;
+struct mipi_dsi_device;
+
+/* Read/write DSI buffer */
+
+struct mipi_dsi_msg
```
Comment (sunghan-chang) [inline os/include/tinyara/mipidsi/mipi_dsi.h:115]: Need to apply TizenRT coding rules.

File: os/drivers/lcd/lcd_dev.c (hunk 6 above)
Comment (namanjain7) [reply os/drivers/lcd/lcd_dev.c:71]: We have defined this variable here: include/tinyara/lcd/lcd_dev.h:34:#define MAX_NO_PLANES 3
This variable was being used in two files: "drivers/lcd/lcd_dev.c" and "drivers/lcd/st7701.c"
So, if we add this in include/tinuara/lcd/lcd_dev.h , then both the files will be able to use this.

File: os/drivers/mipidsi/mipi_dsi.h [hunk 16]
```diff
@@ -0,0 +1,102 @@
+/****************************************************************************
+ *
+ * Copyright 2024 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline os/drivers/mipidsi/mipi_dsi.h:3]: Don't you get it from other opensource?
If you did, you should keep original license and should add Samsung license on top of it.
Could you check this?

**Comment by namanjain7** [comment]
I have fixed the above comments and raised a new PR: https://github.com/Samsung/TizenRT/pull/6148

File: os/drivers/lcd/lcd_dev.c (hunk 6 above)
Comment (namanjain7) [reply os/drivers/lcd/lcd_dev.c:71]: I have defined in lcd_dev.h
path: os/include/tinyara/lcd/lcd_dev.h
I am using this variable in st7701.c driver, and defining again in st7701.c will make it duplicate.

File: os/arch/arm/src/amebasmart/amebasmart_mipi.c [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+		word0 = (cmd_addr[idx + 3] << 24) + (cmd_addr[idx + 2] << 16) + (cmd_addr[idx + 1] << 8) + cmd_addr[idx + 0];
+		word1 = (cmd_addr[idx + 7] << 24) + (cmd_addr[idx + 6] << 16) + (cmd_addr[idx + 5] << 8) + cmd_addr[idx + 4];
+
+		MIPI_DSI_CMD_LongPkt_MemQWordRW(MIPIx, addr, &word0, &word1, FALSE);
+	}
+
+	MIPI_DSI_CMD_Send(MIPIx, MIPI_DSI_DCS_LONG_WRITE, payload_len, 0);
+}
+
+static void MipiDsi_ST7701S_isr(void)
```
[the remark this answers is not in the document]
Comment (namanjain7) [reply os/arch/arm/src/amebasmart/amebasmart_mipi.c:242]: It's because of the name, it's not tightly related. I have changed the name in new PR.

File: os/board/rtl8730e/src/rtl8730e_st7701.c (hunk 5 above)
Comment (namanjain7) [reply os/board/rtl8730e/src/rtl8730e_st7701.c:202]: can you please clarify, what's "no line end" warning.
